### ISSUE TITLE
Appveyor tests failing

### ISSUE THREAD

**Issue by jd-boyd** [civetweb/civetweb#278]
This environment works:
```
Environment: compiler=msvc-18-seh, build_shared=NO, no_files=NO, enable_ipv6=NO, enable_ssl=YES, enable_websockets=YES, no_cgi=NO; Platform: x86
```

This one fails:
```
Environment: compiler=msvc-18-seh, build_shared=NO, no_files=NO, enable_ipv6=NO, enable_ssl=YES, enable_websockets=YES, no_cgi=NO; Platform: x64
```
Note, the only difference is in the failure is in the 64bit windows version.

Here is the summary:
```
  Test project C:/projects/civetweb/output/build
        Start  1: test-private-http-message
   1/19 Test  #1: test-private-http-message ...................   Passed    0.04 sec
        Start  2: test-private-url-parsing
   2/19 Test  #2: test-private-url-parsing ....................   Passed    0.02 sec
        Start  3: test-private-internal-parsing
   3/19 Test  #3: test-private-internal-parsing ...............   Passed    0.00 sec
        Start  4: test-private-encode-decode
   4/19 Test  #4: test-private-encode-decode ..................   Passed    0.01 sec
        Start  5: test-private-mask-data
   5/19 Test  #5: test-private-mask-data ......................   Passed    0.01 sec
        Start  6: test-publicfunc-version
   6/19 Test  #6: test-publicfunc-version .....................   Passed    0.01 sec
        Start  7: test-publicfunc-options
   7/19 Test  #7: test-publicfunc-options .....................   Passed    0.01 sec
        Start  8: test-publicfunc-mime-types
   8/19 Test  #8: test-publicfunc-mime-types ..................   Passed    0.01 sec
        Start  9: test-publicfunc-strcasecmp
   9/19 Test  #9: test-publicfunc-strcasecmp ..................   Passed    0.01 sec
        Start 10: test-publicfunc-url-encoding-decoding
  10/19 Test #10: test-publicfunc-url-encoding-decoding .......   Passed    0.01 sec
        Start 11: test-publicfunc-cookies-and-variables
  11/19 Test #11: test-publicfunc-cookies-and-variables .......   Passed    0.01 sec
        Start 12: test-publicfunc-md5
  12/19 Test #12: test-publicfunc-md5 .........................   Passed    0.00 sec
        Start 13: test-publicserver-check-test-environment
  13/19 Test #13: test-publicserver-check-test-environment ....   Passed    0.02 sec
        Start 14: test-publicserver-start-threads
  14/19 Test #14: test-publicserver-start-threads .............   Passed    1.16 sec
        Start 15: test-publicserver-start-stop-http-server
  15/19 Test #15: test-publicserver-start-stop-http-server ....   Passed    3.25 sec
        Start 16: test-publicserver-start-stop-https-server
  16/19 Test #16: test-publicserver-start-stop-https-server ...***Failed    1.03 sec
        Start 17: test-publicserver-tls-server-client
  17/19 Test #17: test-publicserver-tls-server-client .........***Failed    1.02 sec
        Start 18: test-publicserver-server-requests
  18/19 Test #18: test-publicserver-server-requests ...........***Failed    0.00 sec
        Start 19: test-exe-helper-funcs
  19/19 Test #19: test-exe-helper-funcs .......................   Passed    0.02 sec
  
  84% tests passed, 3 tests failed out of 19
  
  Total Test time (real) =   7.04 sec
```

I think if test-publicserver-start-stop-https-server, test-publicserver-start-stop-https-server, test-publicserver-start-stop-https-server were to be run directly (seperately from CTest), it would give more useful information about the test failure, but CTest is suppressing that information.

An alternative could be to turn on XML logging and have the xml files be uploaded to appveyor as test results.

Here is the link explaining uploading: https://www.appveyor.com/docs/running-tests#uploading-xml-test-results
here is the link explaining how to do xml logging in libcheck: ftp://www.eeng.dcu.ie/pub/ee454/cygwin/usr/share/doc/check-0.9.2/html/x232.html

**Comment by bel2125** [comment]
Thank you for the pull request to turn on logging. Now I have at least an idea what may have gone wrong:
```
C:\projects\civetweb\test\public_server.c:340:F:Start Stop HTTPS Server:test_mg_start_stop_https_server:0: Assertion 'errmsg==""' failed: errmsg=="load_dll: cannot load libeay32.dll", ""=="
```
The reason the 32 bit build works and the 64 bit build does not is most likely `libeay32.dll` is a 32 bit DLL and the 64 bit Windows version must load `libeay64.dll`.

**Comment by bel2125** [comment]
It seems this is not really better:
```
C:\projects\civetweb\test\public_server.c:340:F:Start Stop HTTPS Server:test_mg_start_stop_https_server:0: Assertion 'errmsg==""' failed: errmsg=="load_dll: cannot load libeay64.dll", ""==""
```
I guess I will have to try to get the appveyor script somehow running on a Win64 PC.

**Comment by bel2125** [comment]
I tried to reproduce the Win64 issue locally with Visual Studio 2015 Community, but it seems my 30 days trial period is over. It did not accept any of my emails to register it. Then I created a new hotmail address, failed approximately 10 times at the captcha, and finally it did not accept this email either.

So I tried to reproduce at least the error in the CI test for Linux (CGI fails with error 500 when executing test.cgi in the unit test, although test.cgi should be the same as linux.cgi which works fine in the local test). However, I could not get some travis equivalent running locally, and I did not find out how to get the tests executed by travis run locally either.

Probably just a bad day for fixing the CI test errors ...

Currently the only way I know to debug CI tests is something like
```
sprintf(msg, "Some diagnostic text %s", something); 
ck_abort_msg(msg);
```
which is very annoying and generates a lot of commit noise.
If there would be some way to see the `printf` outputs in the test logs, it would be much easier.

**Comment by jd-boyd** [comment]
A heads up on a possible issue with the 32bit mingw build, which is currently obscured by the current failure:

On line 313 of civetweb.c:
```c
#define timegm(x) (_mkgmtime(x))
```

On some versions of mingw for 32bit windows (4.9.2, 4.8.3), `_mkgmtime32` isn't exposed correctly.  This doesn't effect the same versions for 64bit windows, nor does it effect MSVC.

I'm mentioning this in case it also comes up again when the appveyor tests make it far enough to start doing the mingw tests.

In my own project I added an implementation of that function taken from the musl project.

**Comment by kangjoni76** [comment]
@bel2125 you could try alternative minimalist command line version via nuget.exe, see https://blogs.msdn.microsoft.com/vcblog/2016/02/16/try-out-the-latest-c-compiler-toolset-without-waiting-for-the-next-update-of-visual-studio/

**Comment by jd-boyd** [comment]
> I tried to reproduce the Win64 issue locally with Visual Studio 2015 Community, but it seems my 30 days trial period is over. It did not accept any of my emails to register it. Then I created a new hotmail address, failed approximately 10 times at the captcha, and finally it did not accept this email either.

I thought that Visual Studio 2015 Community was supposed to be free for individuals, per 
[Free Dev Tools - Visual Studio Community 2015
](https://www.visualstudio.com/products/visual-studio-community-vs
) it says:
> Any individual developer can use Visual Studio Community to create their own free or paid apps.

I installed that at work, but I haven't tried running it recently (my own project there is using mingw for now).

> Currently the only way I know to debug CI tests is something like
> 
>``` 
sprintf(msg, "Some diagnostic text %s", something); 
> ck_abort_msg(msg);
```

The way to perhaps stream line that is to first add a lot of that at once.

A second thing that can be done to help is to publish more results as artifacts.  See
[Packaging artifacts - AppVeyor](https://www.appveyor.com/docs/packaging-artifacts
) for details.  Broadly, you can either label artifacts to save in the appveyor.yml file, or you can do so in the script like this: `appveyor PushArtifact <file_name>`

This may be useful a few ways.  One, the build test .exe files could be saved for running locally, second, additional log files could be written and pushed as artifacts.

>  which is very annoying and generates a lot of commit noise.
> If there would be some way to see the printf outputs in the test logs, it would be much easier.

I haven't figured out capturing the stdout/stderr output in a better manner yet, but it is certainly one of the things I'd like to figure out to ease the tests on linux as well as windows.

**Comment by bel2125** [comment]
> I thought that Visual Studio 2015 Community was supposed to be free for individuals

It is - still they want you to register with a Microsoft ID (could be a hotmail account) after a "trial" of 30 days - and this did not work. It's probably just a small detail that made it not work - but if something starts to annoy me I better put it aside for a few days.

> you could try alternative minimalist command line version via nuget.exe

I'm going to try this as well.

> The way to perhaps stream line that is to first add a lot of that at once.

If you mean, I should concat all messages to a long string - this will not work, since the abort message is cut after some letters. At least for ch_string_eq(a, b), if a and b are different, they are truncated to a quite short string, so I can not even use the existing error messages "Error in function ...." since they are truncated before the interesting part of the message begins.

> A second thing that can be done to help is to publish more results as artifacts.

I added some artifacts to the appveyor.yml file, but it does not seem to work in the way I understood the documentation. After build and test worked, [uploading artifacts failed](https://ci.appveyor.com/project/civetweb/civetweb/build/413/job/hxoxhualuwg0j9hw#L1036):
`Packaging artifacts...Pattern contains invalid characters.`
I probably need to get to a good appveyor file by try and error again.

**Comment by bel2125** [comment]
The OSX tests currently seem to fail because of the check framework:
https://travis-ci.org/civetweb/civetweb/jobs/117785296

**Comment by bel2125** [comment]
Appveyor currently fails with build script errors. I do not understand the exact problem, and have no idea how to solve it.

**Comment by jd-boyd** [comment]
On the first job I read of the most recent build the problem is that the patch command I added (for a fix for osx) to the cmake script  is failing  because there is no patch command (or it just isn't in the path) on windows.  This should be easy to fix.

**Comment by bel2125** [comment]
On my PC, patch is in git\bin\patch.exe - probably there is no git installed.
Do you think I should install git from https://git-scm.com/download/win in the install step of appveyor.yml?

**Comment by bel2125** [comment]
Checking appveyor.yml again, I think installing another git is not the solution.
There is already a git installed, but it is explicitly removed from the path in appveyor.yml:
```
before_build:
  # Remove sh.exe from the path otherwise CMake will complain:
  # "sh.exe was found in your PATH, here: C:/Program Files/Git/usr/bin/sh.exe"
  # and the MinGW build will not work (the Visual Studio build does not care).
  # See http://help.appveyor.com/discussions/problems/3193-cmake-building-for-mingw-issue-with-git-shexe
  - set PATH=%PATH:C:\Program Files\Git\usr\bin;=%
```
probably it should be left in the path but sh.exe should be renamed.

**Comment by bel2125** [comment]
The fix above seems to work.

Still, there is a 64 bit issue for Visual studio: "load_dll: cannot load libeay64.dll"

And there is the MinGW issue already mentioned above: "undefined reference to `_mkgmtime32'".
But this seems to be a MinGW bug: https://sourceforge.net/p/mingw-w64/bugs/473/

**Comment by kainjow** [comment]
That only fixes it for Travis. If somehow you don't have patch on your system, the Windows build will still fail. I think a proper fix is to only apply the patch file on OS X, since that's only where it's needed. Should be easy, I'll look into a fix.

Also, should move to using CMake for the Appveyor builds, then you can remove the VS project files.

**Comment by jd-boyd** [comment]
I basically agree, but I would modify that to apply the patch when not on Windows. I had started a branch to take that approach to solving the patch problem, but I didn't make a pull request because I saw that bel2125 beat me too it with a different approach.

BTW, Atlassian SourceTree appears to include `patch.exe` here: `C:\Program Files (x86)\Atlassian\SourceTree\tools`

An Atlassian help post said that the entire git toolset could be found: `%USERPROFILE%\AppData\Local\Atlassian\SourceTree\git_local\bin`

**Comment by kainjow** [comment]
I think my existing pull request should solve it: https://github.com/civetweb/civetweb/pull/294

**Comment by jd-boyd** [comment]
> And there is the MinGW issue already mentioned above: "undefined reference to `_mkgmtime32'".
> But this seems to be a MinGW bug: https://sourceforge.net/p/mingw-w64/bugs/473/

I got around that in my own application by copying an implementation of `mkgmtime` from the musl project (also MIT license) and only including it in the build when building with mingw.  That seems fairly dirty though.  I don't know mingw well enough to know if there would be a way to expose that function with a custom `.lib` file.

**Comment by bel2125** [comment]
Regarding the build servers, I did not intend to beat anyone into any direction - I think you (all) have more experience with this build server. So if you prepare a pull request to change the build scripts, I will trust you did the right thing right and merge it.

#293 and #294 are merged. Thank you.

**Comment by bel2125** [comment]
Currently, all AppVeyor tests fail during the build setup with a message like
```
C:\ssl\Win32OpenSSL_102h.exe /SILENT /LOG="C:\ssl\install_ssl32.log" && exit 0
The system cannot execute the specified program.
Command exited with code 1
```
I have no idea what to do about it. I even added the `&& exit 0` according to some documentation I found somewhere in the internet, so the command must return exit code 0 (not 1), but this does not seem to work.

I would like to have AppVeyor fixed before a version 1.8 release, but I don't know how to.
Anyone any idea?

**Comment by kainjow** [comment]
Isn't && a *nix shell thing? Doesn't cmd on Windows just use &?

**Comment by bel2125** [comment]
The && executed two commands:
```
echo X && echo Y
```
Prints X and Y in two lines in cmd.

Beyond believe: it works on Windows CMD and Linux (BA)SH ;-)

**Comment by bel2125** [comment]
I got the && from some AppVeyor discussion group (I do not remember the link).
But, according to the [Microsoft documentation](https://www.microsoft.com/resources/documentation/windows/xp/all/proddocs/en-us/ntcmds_shelloverview.mspx?mfr=true), the && execute the second command only if the first one was successful. So the comment on the AppVeyor discussion group was wrong, and one should use &.
I will give it a try.

**Comment by kainjow** [comment]
Looks like that got it to proceed, but now tests are failing on all platforms :)

**Comment by bel2125** [comment]
This was not the only commit in the last push.
I added ~4 new tests, some of them have warnings in the unit test code (clang only), and warnings are considered as errors. I think it's OK to consider warnings as errors in `civetweb.c`, while it seems somewhat excessively nit-picky to stop the build for a signed / unsigned warning in the test code - anyway, I'm going to fix this warnings.

When the new tests work, let's see what is left with AppVeyor.
One problem was this one: https://github.com/civetweb/civetweb/commit/ca0944f3697eb5b70d005bdc90e72d421e5c0ed5
Let's see what is left when the tests work again.

**Comment by bel2125** [comment]
The new tests for Linux and Win32 are fixed.
Due to the new tests, the test coverage increased from 62% to 65%.
It seems the coverity badge is red below 70% - I tried to change this configuration before, but it did not work.

Some AppVeyor tests work now:
https://ci.appveyor.com/project/civetweb/civetweb/build/499
But it seems there are some problems with OpenSSL on Win64 test machines.
https://ci.appveyor.com/project/civetweb/civetweb/build/499/job/jwjs6rpo7lsb77po#L1131

**Comment by bel2125** [comment]
OpenSSL should already be installed, according to 
 - http://help.appveyor.com/discussions/questions/1132-openssl-installation-issues
 - https://github.com/appveyor/ci/issues/576
Probably `OPENSSL_ROOT_DIR` needs to be set to `C:\OpenSSL-Win32` or `C:\OpenSSL-Win64` and added to the path, so the required DLLs can be found there.

**Comment by bel2125** [comment]
I found a somewhat ugly way to get HTTPS tests with OpenSSL on Win64 AppVeyor build images working: It seems both, OpenSSL 32 and 64 bit are already downloaded and installed into C:\\OpenSSL-Win32\\ and C:\\OpenSSL-Win64\\ in the AppVeyor build images. It seems the 64 bit libraries are called libeay32.dll and libssl32.dll, just like the 32 bit DLLs. I got the tests working by copying and renaming the DLLs there into the local directory: https://github.com/civetweb/civetweb/commit/23f8726eed78dfd564579801700f6441c241235d
Ugly, but works.
If someone knows something less ugly, please tell me - or better prepare a pull request. Otherwise I will just leave it like this and close this issue.

Not, most but not all of the builds are working:
https://ci.appveyor.com/project/civetweb/civetweb/build/521

The remaining issue:
https://ci.appveyor.com/project/civetweb/civetweb/build/521/job/3xx2f9ua25sb4and#L187
```
CMake Error at C:/Program Files (x86)/CMake/share/cmake-3.5/Modules/CMakeTestCCompiler.cmake:61 (message):
  The C compiler "C:/MinGW/bin/gcc.exe" is not able to compile a simple test
  program.
 
  It fails with the following output:
 
   Change Dir: C:/projects/civetweb/output/build/CMakeFiles/CMakeTmp 
 ```
... whatever this means.

I think we have a 32 bit and a 64 bit build, both with Visual Studio and MinGW - so I will probably just eliminate these two configurations, instead of doing original research on curious error messages.
Or, does someone know what to do here?

**Comment by bel2125** [comment]
I think these four builds are exactly the same:
[Build 521](https://ci.appveyor.com/project/civetweb/civetweb/build/521) ("Try to correct build scripts") - the 12th and 16th configuration fail
[Build 522](https://ci.appveyor.com/project/civetweb/civetweb/build/522) ("Auto format code after merge") - 12 fails, others work including 16
[Build 523](https://ci.appveyor.com/project/civetweb/civetweb/build/523) ("Change nothing - see how it affects the AppVeyor build") - 12 fails, like for 522
[Build 524](https://ci.appveyor.com/project/civetweb/civetweb/build/524) ("Change nothing again - see how it affects the AppVeyor build") - everything works

Nothing of relevance is changed between them. From 521 to 522, some indention has been changed. From 522 to 523 and from 523 to 524, I added and removed an empty line and spaces at the end of the appveyor.yml file. 

I start to believe this is an issue of the AppVeyor service now, no longer of CivetWeb.

I will do a rebuild of 523, to see if the result is the same again:
[Build 525](https://ci.appveyor.com/project/civetweb/civetweb/build/525) - (Rebuild of "Change nothing - see how it affects the AppVeyor build")

And also a rebuild of 522:
[Build 526](https://ci.appveyor.com/project/civetweb/civetweb/build/526) - (Rebuild of "Auto format code after merge")

Build history: https://ci.appveyor.com/project/civetweb/civetweb/history

**Comment by bel2125** [comment]
For unknown reasons, the AppVeyor build for 525 and 526 completed without errors.
Maybe they had some stability issue with the build images ... let's observe this for a few more commits/pushes. If it does not occur again, I will close this issue.

**Comment by bel2125** [comment]
[Build 527](https://ci.appveyor.com/project/civetweb/civetweb/build/527) - configurations 9 + 12 fail
[Build 528](https://ci.appveyor.com/project/civetweb/civetweb/build/528) - all OK
[Build 529](https://ci.appveyor.com/project/civetweb/civetweb/build/529) - 11 fails
rebuild all of them
[Build 530](https://ci.appveyor.com/project/civetweb/civetweb/build/530) (rebuild of 527) - 13 fails
[Build 531](https://ci.appveyor.com/project/civetweb/civetweb/build/531) (rebuild of 528)  - all OK
[Build 532](https://ci.appveyor.com/project/civetweb/civetweb/build/532) (rebuild of 529)  - all OK
:confused:

[Build 534](https://ci.appveyor.com/project/civetweb/civetweb/build/534) (2nd rebuild of 527) - 10 fails
[Build 535](https://ci.appveyor.com/project/civetweb/civetweb/build/535) (3rd rebuild of 527) - all OK
It seems one can repeat any build until it works. :confused:  :question: :confused:

I don't think this is a problem in civetweb anymore - it's probably an issue of the AppVeyor build worker images. I can not definitely exclude that some configuration in civetweb could be set to avoid the problem, but there is absolutely no hint which one. From my point of view, now this issue would be better on the AppVeyor issue tracker than on civetweb.

This was one of the issues I wanted to be solved before the 1.8 release (#265). I think now it is no longer a reason to postpone the release any longer.
